src/gpu/transfomer.ts

```diff
@@ -0,0 +1,321 @@
+          nx,
+          create.memberExpression(create.identifier('Math'), term),
+          args
+        )
+      }
+    })
+
+    // 6. Update all external variable references in body
+    // e.g. let res = 1 + y; where y is an external variable
+    // becomes let res = 1 + this.constants.y;
+
+    const names = [this.outputArray.name, ...this.counters, 'Math']
+    simple(this.targetBody, {
+      Identifier(nx: es.Identifier) {
+        // ignore these names
+        if (names.includes(nx.name) || locals.has(nx.name)) {
+          return
+        }
+
+        create.mutateToMemberExpression(
+          nx,
+          create.memberExpression(create.identifier('this'), 'constants'),
+          create.identifier(nx.name)
+        )
+      }
+    })
+
+    // 7. Update reference to counters
+    // e.g. let res = 1 + i; where i is a counter
+    // becomes let res = 1 + this.thread.x;
+
+    // depending on state the mappings will change
+    let threads = ['x']
+    if (this.state === 2) threads = ['y', 'x']
+    if (this.state === 3) threads = ['z', 'y', 'x']
+
+    const counters: string[] = []
+    for (let i = 0; i < this.state; i = i + 1) {
+      counters.push(this.counters[i])
+    }
+
+    simple(this.targetBody, {
+      Identifier(nx: es.Identifier) {
+        let x = -1
+        for (let i = 0; i < counters.length; i = i + 1) {
+          if (nx.name === counters[i]) {
+            x = i
+            break
+          }
+        }
+
+        if (x === -1) {
+          return
+        }
+
+        const id = threads[x]
+        create.mutateToMemberExpression(
+          nx,
+          create.memberExpression(create.identifier('this'), 'thread'),
+          create.identifier(id)
+        )
+      }
+    })
+
+    // 8. we transpile the loop to a function call, __createKernel
+    const kernelFunction = create.functionExpression([], this.targetBody)
+    create.mutateToExpressionStatement(
+      node,
+      create.callExpression(
+        GPUTransformer.globalIds.__createKernel,
+        [
+          create.arrayExpression(this.end),
+          create.objectExpression(externObject),
+          kernelFunction,
+          this.outputArray,
+          tempNode
+        ],
+        node.loc!
+      )
+    )
+
+    return this.state
+  }
+
+  // verification of outer loops using our verifier
+  checkOuterLoops = (node: es.ForStatement) => {
+    let currForLoop = node
+    while (currForLoop.type === 'ForStatement') {
+      const detector = new GPULoopVerifier(currForLoop)
+      if (!detector.ok) {
+        break
+      }
+
+      this.innerBody = currForLoop.body
+      this.counters.push(detector.counter)
+      this.end.push(detector.end)
+
+      if (this.innerBody.type !== 'BlockStatement') {
+        break
+      }
+
+      if (this.innerBody.body.length > 1) {
+        break
+      }
+
+      currForLoop = this.innerBody.body[0]
+    }
+  }
+
+  /*
+   * Based on state, gets the correct body to me run across threads
```

Comment:
nitpicking: looks like 'me' should be 'be'

Reply:
yup that's right!